include/rcpputils/patch.hpp

```diff
@@ -0,0 +1,200 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCPPUTILS__PATCH_HPP_
+#define RCPPUTILS__PATCH_HPP_
+
+#include <functional>
+#include <string>
+#include <utility>
+
+#include "mimick/mimick.h"
+#include "rcutils/macros.h"
+
+namespace rcpputils
+{
+
+template<size_t N, typename SignatureT>
+struct patch_traits;
```

Comment:
For these structs, c++ style suggests using PascalCase yah?

Reply:
Fair enough. I tend to fallback to `stl` style whenever writing templates.